llvm/tools/sycl-post-link/SpecConstants.cpp

```diff
@@ -300,26 +293,35 @@ Constant *getConstantElement(Constant *C, unsigned i, Type *elementType) {
 void collectCompositeElementsDefaultValuesRecursive(
     const Module &M, Constant *C, unsigned &Offset,
     std::vector<char> &DefaultValues) {
-  Type *Ty = C->getType();
-  if (auto *DataSeqC = dyn_cast<ConstantDataSequential>(C)) {
+  if (isa<ConstantAggregateZero>(C)) {
+    // This code is generic for zeroinitializer for both arrays and structs
+    size_t NumBytes = M.getDataLayout().getTypeStoreSize(C->getType());
+    std::fill_n(std::back_inserter(DefaultValues), NumBytes, 0);
+    Offset += NumBytes;
+  } else if (auto *DataSeqC = dyn_cast<ConstantDataSequential>(C)) {
     // This code is generic for both vectors and arrays of scalars
     for (size_t I = 0; I < DataSeqC->getNumElements(); ++I) {
       Constant *El = cast<Constant>(DataSeqC->getElementAsConstant(I));
       collectCompositeElementsDefaultValuesRecursive(M, El, Offset,
                                                      DefaultValues);
     }
-  } else if (auto *ArrTy = dyn_cast<ArrayType>(Ty)) {
+  } else if (auto *ArrayC = dyn_cast<ConstantArray>(C)) {
     // This branch handles arrays of composite types (structs, arrays, etc.)
+    auto *ArrTy = ArrayC->getType();
     for (size_t I = 0; I < ArrTy->getNumElements(); ++I) {
-      Constant *El = getConstantElement(C, I, ArrTy->getElementType());
+      // We are shure the array is not a zeroinitializer.
+      Constant *El = cast<Constant>(C->getOperand(I));
```

Comment:
```suggestion
      Constant *El = ArrayC->getOperand(I);
```
and remove `ArrTy` declaration.

Or you can pass `ArrayC->getOperand(I)` directly to `collectCompositeElementsDefaultValuesRecursive` avoiding local variable.